spec/immutable-data-structures.html

```diff
@@ -6,27 +6,590 @@ <h1>Immutable Data Structures</h1>
+          1. Let _fields_ be a new empty List.
+          1. For each element _prop_ of _props_,
+            1. Let _name_ be ! GetV(_prop_, `"0"`).
+            1. Let _value_ be ! GetV(_prop_, `"1"`).
+            1. If Type(_value_) is Object, throw a *TypeError* exception.
+            1. Let _field_ be { [[Key]]: _name_, [[Value]]: _value_ }.
+            1. Append _field_ to the end of list _fields_.
+          1. Return a new Record value whose [[Fields]] is _fields_.
+        </emu-alg>
+      </emu-clause>
+    </emu-clause>
+    <emu-clause id="sec-properties-of-the-record-constructor">
+      <h1>Properties of the Record Constructor</h1>
+      <p>The Record constructor:</p>
+      <ul>
+        <li>has a [[Prototype]] internal slot whose value is %Record.prototype%.</li>
+        <li>has the following properties:</li>
+      </ul>
+      <emu-clause id="sec-record.fromentries">
+        <h1>Record.fromEntries ( _iterable_ )</h1>
+        <p>The *fromEntries* function takes one argument _iterable_, and performs the following steps:</p>
+        <emu-alg>
+          1. Perform ? RequireObjectCoercible(_iterable_).
+          1. Let _fields_ be a new empty List.
+          1. Let _iteratorRecord_ be ? GetIterator(_iterable_).
+          1. Let _stepsDefine_ be the algorithm steps defined in <emu-xref href="#sec-append-record-field-to-list-functions" title></emu-xref>.
+          1. Let _adder_ be ! CreateBuiltinFunction(_stepsDefine_, &laquo; &raquo;).
+          1. Perform ! AddEntriesFromIterable(_fields_, _iterable_, _adder_).
+          1. Return a new Record value whose [[Fields]] is _fields_.
+        </emu-alg>
+        <emu-note>
+          <p>The parameter _iterable_ is expected to be an object that implements an @@iterator method that returns an iterator object that produces a two element array-like object whose first element is a value that will be used as a Map key and whose second element is the value to associate with that key.</p>
+        </emu-note>
+      </emu-clause>
+      <emu-clause id="sec-append-record-field-to-list-functions">
+        <h1>AppendRecordFieldToList Functions</h1>
+        <p>An AppendRecordFieldToList function is an anonymous built-in function. When an AppendRecordFieldtoList function is called with arguments _key_ and _value_, the following steps are taken:</p>
+        <emu-alg>
+            1. Let _list_ be the *this* value.
+            1. Assert: Type(_list_) is List.
+            1. If Type(_value_) is Object, throw a *TypeError* exception.
+            1. Let _field_ be { [[Key]]: _key_, [[Value]]: _value_ }.
+            1. Append _field_ to the end of list _list_.
+        </emu-alg>
+      </emu-clause>
+      <emu-clause id="sec-record.isrecord">
+        <h1>Record.isRecord ( _arg_ )</h1>
+        <p>The *isRecord* function takes one argument _arg_, and performs the following steps:</p>
+        <emu-alg>
+          1. Return ? IsRecord(_arg_).

```

Comment:
Where is this defined, and when could it return an abrupt completion?

Reply:
Probably this could be an ! ; note that there are multiple parallel PRs in development, so another one which covers the data model could backfill the definition of IsRecord.

Early reviews can be good, but let's focus on offering constructive suggestions to newer spec authors. There will be lots of time to pick things apart for remaining errors after the champion group and collaborators collectively write this text.

At the same time, I think we're far along enough in the development of the data model that we can fill in concrete text here. I think the `IsRecord` algorithm will not have any other uses, so we can describe this concretely. It could look something like this:

1. If Type(_arg_) is Record, return *true*.
1. If Type(_arg_) is Object and _arg_ has a [[RecordData]] internal slot, return *true*.
1. Return *false*.

Similar wording could be used for Tuple.isTuple.

Reply:
I find all of my suggestions constructive; you’re free to apply your own rubric to your own suggestions.

Reply:
To clarify, I am really thankful for all of the comments from @ljharb and @littledan, especially as a brand-new spec author. I'm sure there is a massive list of work left to do here, and am eager to soak up as much info as possible.